### ISSUE TITLE
Fully fledged #includes instead of forward-declared classes are generated

### ISSUE THREAD

**Issue by vi** [dropbox/djinni#165]
For the second sample from #164, Djinni generates the following header file `toxml.hpp`:

```c++
// AUTOGENERATED FILE - DO NOT MODIFY!
// This file generated by Djinni from test.djinni

#pragma once

#include "anotherrecord.hpp"
#include "somerecord.hpp"
#include <memory>
#include <string>

namespace qqq {

class Toxml {
public:
    virtual ~Toxml() {}

    virtual std::string toxml() = 0;

    static std::shared_ptr<Toxml> castFromSomeRecord(const Somerecord & sr);

    static std::shared_ptr<Toxml> castFromAnotherRecord(const Anotherrecord & ar);
};

}  // namespace qqq

```

This causes cirtual header inclusing and doesn't build.

I think Djinji should generate forward class declarations instead, like this:

```
class Somerecord;
class Anotherrecord;
```

With this change it builds and successfully runs.

**Comment by artwyman** [comment]
I agree that a forward-declaration would be sufficient here, so I'd be in favor of changing this.  The use case you're creating is a bit outside the expectations of Djinni, which is why we haven't run into this before.  @vi do you want to try to implement this change?  Pull requests are welcome.

**Comment by vi** [comment]
> Do you want to try to implement this change?

I can try preparing a pull request implementing a naive change from `#include ...` to `class ...`. I can't run tests although.

```
make test
make -C test-suite
Already up to date: Djinni
...
cd objc; xcodebuild -sdk iphonesimulator -project DjinniObjcTest.xcodeproj -scheme DjinniObjcTest test
/bin/sh: 1: xcodebuild: not found
```

That casts some Apple vibes, scaring away from using tests...

**Comment by artwyman** [comment]
What platform are you working with?  Most of the Djinni build/test frameworks do assume you're running on a Mac.  Some parts of it will work on Linux, but not all.  We could do better at reducing that, but realistically if we want to support ObjC for iOS you need to be on a Mac anyway.

**Comment by vi** [comment]
I also tried running "make java" in `test-suite`, but it has also failed:

```
    [javac] /mnt/src/git/djinni/support-lib/java/com/dropbox/djinni/NativeLibLoader.java:92: ')' expected
    [javac]             for (Path p : (Iterable<Path>)streamPaths::iterator) {
```

**Comment by artwyman** [comment]
That doesn't sound like a platform problem, and that test suite should work.  Do you see the same problem with unmodified Djinni code?  What's your JDK version?

**Comment by vi** [comment]
That's unmodified Djinni, commit 946e0ac504ccc84ac5eff427059593caf0f17520. Unfortunately modified fails earlier with error probably related to my changes.

Java is probably java-7-openjdk-i386.

**Comment by artwyman** [comment]
Ah, looks like some Java-8-only code got introduced in commit @27e29479.  Can you upgrade to a Java 8 JDK?  We might want to reconsider whether Djinni's tests should depend on Java 8, but upgrading should get you unblocked for now.

**Comment by fostah** [comment]
I'm currently updating my app to use the latest version of Djinni and ran into a compiler error because of this change. Here's a strip downed example of the issue:

We have an enum, called Type:

# List of available types
Type = enum {
    Type1;
    Type2;
}

We import Type into a Djinni manager, and use it as the key of a map:

@import "Type.djinni"
@import "TypeObject.djinni" // just assume this exists...

# A Manager...
Manager = interface +c {
    # Initialize the Manager
    static initializeManager(types: map<Type, TypeObject>): Manager;    
}

Because of forward declaring, the generated C++ for Manager no longer compiles. We get the following error:

**Error - implicit instantiation of undefined template**

When you dig deeper, the issue is because the unordered_map that's generated for map does not have the entire definition of the Type enum.

**Comment by vi** [comment]
Shall enums be special-cased to include full `#include `? Or is the problem not specific to enums and can occur with other types as well?

**Comment by fostah** [comment]
I think having a special-case for enums is all that's needed. 

Classes as map keys seem to compile fine, probably because they're shared_ptrs. And structs as map keys don't seem to compile in either case, because there's no hash function defined for the Djinni generated structs.

**Comment by artwyman** [comment]
@fostah I think I might be missing a detail, since I just tried adding the declarations you gave to the example app's example.djinni, and the result compiled (after I provided a cpp file which defines initializeManager to return nullptr).  Can you hack up the example into something which demonstrates the failure?  My suspicion is that the full definition isn't needed in the header, but it would be needed in cpp files which use the method in question in certain situations.  For hand-written code you can handle that, but it's possible that the generator might need to add the #include in some scenarios which it isn't yet.

**Comment by fostah** [comment]
@artwyman Sure. Here's how I edited the example project. To simplify things, I just added a new static method to sort_items:

```
sort_items = interface +c {
    static test(sortMap: map<sort_order, textbox_listener>);
    // ...
}
```
Then in sort_items_impl I just have a blank implementation of test:

```
void SortItems::test(const std::unordered_map<sort_order, std::shared_ptr<TextboxListener>> & sortMap) {
    
}
```

When I try to compile (make example_ios), I see the following error:

```
In file included from /Users/mfoster/Workspace/TEMP/dropbox/djinni/example/generated-src/cpp/sort_items.hpp:6:
In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/memory:597:
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/type_traits:964:38: error: implicit instantiation of undefined template 'std::__1::hash<textsort::sort_order>'
    : public integral_constant<bool, __is_empty(_Tp)> {};
```

Note that if I change the key to a string, everything compiles fine.

I've attached the entire build log for you to take a closer look at.

I appreciate the help.

[make_example_ios_log.txt](https://github.com/dropbox/djinni/files/158679/make_example_ios_log.txt)

**Comment by artwyman** [comment]
Thanks for those specifics.  My build still passed with them, which lead me to discover that what's different is actually a difference between our internal version of the Djinni example and the public version.  Specifically, it looks like the error you're encountering doesn't occur if building with C++14, which we do internally, and made the default in the experimental python branch.  The public example uses C++11 by default.  You can confirm this if you change the two instances of "c++11" in common.gypi to be "c++1y instead".  That will make your build succeed.  If that's something you can do for your real builds, then go for it.  C++14 is awesome anyway.  But it should also be possible to make C++11 work.

That plus the specifics of the error makes me realize that what's missing isn't the definition of the enum, but the definition of the specialized std::hash in the enum's header.  It seems that in C++14 the STL already can hash enums by default, without help?  I think that rings a bell as something I heard about in the past.

Regardless, I think the place where the definition is needed isn't necessarily in the sort_items.hpp header, but in TXSSortItems+Private.mm where the map actually needs to be created.  It might make sense to generate it there, or in one of the +Private header files for ObjC++.  I'll look into that.

**Comment by artwyman** [comment]
I just submitted new issue #213 to capture this, and I copied some of the details of this discussion over there.

**Comment by fostah** [comment]
Great, thanks. And I'm going to tinker around with C++14 too.
